Add clear button to text inputs

derekblank opened the pull request:
## Description
Addresses [wordpress-mobile/gutenberg-mobile#3017](https://github.com/wordpress-mobile/gutenberg-mobile/issues/3017) by adding clear button behavior to certain TextInput controls within BottomSheet.  See [#3017](https://github.com/wordpress-mobile/gutenberg-mobile/issues/3017) for further discussion on how text input behavior may be standardized.

## How has this been tested?

### Test Case 1: Display the Clear Button and Clear Text Input
- Open the app with metro running.
- Add a **Social Icons** block.
- Tap in to a **Social Icon** to invoke the **Link Settings** modal.
- Tap "URL" or "Link label" to focus the Text Input.
- Add text to the Text Input via the keyboard.
- _Expect_: A clear button should appear if text value has been entered.
- If text has been entered and the clear button is displayed, **tap the Clear Button**.
- _Expect_: Any text input value should disappear, and placeholder text should be shown; Clear Button should not be visible.

### Test Case 2: Do not display the Clear Button for Plain Text Editor components
- Open the app with metro running.
- Add a **Cover** block.
- Tap into the **Cover** caption Text Input to add a caption or text.
- _Expect_: A clear button **should not** appear if text value has been entered.


## Screenshots 

**Android, dark mode:**

https://user-images.githubusercontent.com/643285/146175802-eb6ea8c8-d01c-4ba4-99bf-b45012258caf.mov

**iOS, light mode:**

https://user-images.githubusercontent.com/643285/146175834-2129d1a8-997e-431e-8474-9577851ba6c6.mov




## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

derekblank committed:
Add clear button behavior to BottomSheet.Cell

derekblank committed:
Remove clear button behavior from LinkPicker

derekblank committed:
Update BottomSheet.Cell code style

derekblank committed:
Add displayClearButton prop to Link Settings component

derekblank committed:
Update Link Settings code style

fluiddot commented on packages/components/src/mobile/bottom-sheet/cell.native.js:
```diff
@@ -240,24 +265,27 @@ class BottomSheetCell extends Component {
+						keyboardType={ this.typeToKeyboardType( type, step ) }
+						{ ...valueProps }
+					/>
+					{ displayClearButton && clearButton() }
```
The `clearButton` function checks whether the local state `localValue` is empty to display the button, I'm wondering if we should include that condition here instead of within the function, wdyt?

https://github.com/WordPress/gutenberg/blob/0f6c1941d06d0095409ba8294089ef79427e6ea8/packages/components/src/mobile/bottom-sheet/cell.native.js#L205

derekblank replied:
Good call, updated!

fluiddot commented on packages/components/src/mobile/bottom-sheet/cell.native.js:
```diff
@@ -35,6 +35,7 @@ class BottomSheetCell extends Component {
 		this.state = {
 			isEditingValue: props.autoFocus || false,
 			isScreenReaderEnabled: false,
+			localValue: props.value ? props.value : '',
```
I think you could use here the [Nullish coalescing operator (`??`)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator) to simplify the code.

derekblank replied:
This line is no longer relevant, but this is a good callout about using a better logic operator nonetheless. 👍

fluiddot commented on packages/components/src/mobile/bottom-sheet/cell.native.js:
```diff
@@ -192,6 +194,29 @@ class BottomSheetCell extends Component {
+
+		const clearButton = () => {
+			return (
+				<>
```
I'm wondering if we need to use a fragment here since we're only returning a single instance 🤔 .

derekblank replied:
I think my reasoning here was that in previous projects, the code styles had dictated that you could not return a conditional rendering block using `{ ... }` without using a Fragment as the parent of the curly braced block, so perhaps this was habit. This has been removed, however.

fluiddot replied:
Oh, that's interesting, I was also accustomed to doing something similar in these cases to prevent returning a `false` value, in case the condition isn't fulfilled, instead of an instance, which might produce render issues on Android. I see in the recent changes that this is no longer an issue, so I'm marking this comment as resolved 👍 .

fluiddot commented on packages/components/src/mobile/link-picker/index.native.js:
```diff
@@ -73,20 +73,11 @@ export const LinkPicker = ( {
 	};
 
-	const clear = () => {
-		setValue( { value: '', clipboardUrl } );
```
I noticed that we would stop setting the `clipboardUrl` state value when clearing the text input if we remove this function. Not sure about potential side effects but probably would be worth investigating it.

derekblank replied:
This is a very good point, and an edge case that I was considering how to address. I've updated BottomSheet.Cell to use an onClear handler as a prop, so that the parent component (like this one) can set the TextInput value directly, which handles this case. (More info in my comment below.)

fluiddot replied:
Great, thanks for addressing the issue, I'll check the comments.

fluiddot commented on packages/components/src/mobile/link-picker/styles.native.scss:
```diff
@@ -16,18 +16,6 @@
 }
 
-.clearIcon {
-	margin-left: 16;
```
Not sure if it's needed but I'm wondering if we should carry this style to the `cellTextInputContainer`, wdyt?

fluiddot replied:
As you can see in the following screenshot, the close button looks too close to the text input which could lead users to tap it when trying to interact with the text caret, so we might consider giving extra padding.

<img src=https://user-images.githubusercontent.com/14905380/146238410-6c101f2a-8026-4fac-bff2-4f87136c84cb.png width=400>

derekblank replied:
I think this is a good idea. I've updated the styles to match.

fluiddot replied:
When doing some testing I cleared the input a couple of times by mistake when trying to select text, I noticed that the bounds of the clear button might interfere with the text input itself. This is not an issue when the text is aligned to the left but can be a major problem when is aligned to right. I'm wondering if we could tweak the button to prevent this behavior, wdyt?

<img width="495" alt="Screenshot 2021-12-16 at 20 10 59" src="https://user-images.githubusercontent.com/14905380/146433679-c2c02c8d-ec6f-422b-a4d2-7f733e493528.png">

derekblank replied:
I think this is a good callout, and I've updated the behavior to be more compatible with right aligned text by transferring the margin property from the icon to the icon's touchable container. Initially, I had pulled in the style directly from LinkPicker, which was `margin-left: 16;`, and that felt a bit too wide for right aligned text. Now, after interacting with it a bit, it feels like a good amount, and provides enough buffer space for the user to cleanly interact with both the text and the button, without interference. Thanks for the suggestion, much appreciated. 👍

fluiddot replied:
Awesome, I'll perform another test on this part as a double-check before solving the issue, thanks for addressing it 🙇 .

fluiddot replied:
I've just tested it and I can confirm that the clear button no longer produces this issue 🎊 .

fluiddot reviewed (commented):
Thank @derekblank for tackling this issue 🙇 !

Here are some comments gathered after doing a first review on the draft PR. On the other hand, I noticed that unit tests are failing, so we might consider investigating the culprit of the errors, in case it uncovers any potential issues from the changes.

derekblank committed:
Use onClear handler in BottomSheet.Cell

derekblank committed:
Update BottomSheet.Cell clear icon styles

derekblank committed:
Remove unused localValue state from BottomSheet.Cell

derekblank commented:
Thanks for the feedback @fluiddot ! After considering your comments, and thinking through how we might resolve some of the edge cases like retaining the `clipboardUrl` in LinkPicker, I realized that a better approach would be to provide an `onClear` method as a prop to BottomSheet.Cell. This way, each use of `displayClearButton` in BottomSheet.Cell would be accompanied by its own `onClear` handler that the clear button itself could invoke, so that the TextInput value can be handled directly in the parent component of wherever BottomSheet.Cell is being used:

```
const clearInput = () => {
    setText( '' );
};

<BottomSheet.Cell
    displayClearButton={true}
    onClear={() => clearInput()}
    value={text}
    ...
/>
```

It is an extra prop and function, yes, but I think it is much cleaner as it eliminates the need for a `localValue` state within BottomSheet.Cell, and it allows the parent component to maintain control of the value and provide extra parameters like `setValue( { value, clipboardUrl: '' } )` when needed. In addition, this change resolves several of your other comments on the code, but that feedback was helpful. Let me know your thoughts on this approach! 🙇

fluiddot commented on packages/components/src/mobile/bottom-sheet/cell.native.js:
```diff
@@ -340,10 +331,6 @@ class BottomSheetCell extends Component {
-		const clearButtonIconStyle = getStylesFromColorScheme(
-			styles.clearButtonIcon,
-			styles.clearButtonIconDark
-		);
```
I noticed the clear icon's color is now using the values from `iconStyleBase` instead of having a specific one as we had with `clearButtonIconStyle`, it would be great if we could double-check that those colors are the good ones.

derekblank replied:
I agree with you here that we may wish to have more fine-grained control over the clear button icon color. This code change here was actually an attempt to consolidate color variables a bit, and reduce some of the styling complexity as the colors for `icon` and `iconDark` were already very similar to the new clear button icon colors (which I had brought from LinkPicker directly). With regard to which colors are the "right" ones, there seems to be a bit of variance throughout the app with matching the true native platform UI (if that is even the desired goal, of course). I did note that there is a bit of a difference between the iOS native clear button color, and how we are implementing it. Here are some examples of BottomSheet.Cell usage (top) and Search Control (bottom). In both screenshots, the native iOS clear button (using [`clearButtonMode`](https://developer.apple.com/documentation/uikit/uitextfield/1619622-clearbuttonmode)) is on the **left**, and the Javascript button is on the **right**:

#### **Link Settings**
Link Settings is an example of the BottomSheet.Cell clear button behavior introduced in this PR. (Native iOS clear button on the left, Javascript button on the right using the [existing `icon` and `iconDark` colors](https://github.com/WordPress/gutenberg/blob/2a4257a3b0aaa7c9554cc59d161b90c433b68501/packages/components/src/mobile/bottom-sheet/styles.native.scss#L145-L151).)

![Artboard](https://user-images.githubusercontent.com/643285/146504206-3b472ada-54e1-436d-a917-2769b20e5630.png)

#### **Search Control**
Search Control implements a different clear button solution, using similar colors. (Native iOS clear button is on the left, Javascript on the right.)

![Artboard](https://user-images.githubusercontent.com/643285/146504520-5f98a23f-436a-4e62-ab1e-d83563f821fb.jpg)

These color differences are more noticeable on iOS because the icon has more surface area, but the result is also very similar on Android. While none of the colors in the examples seem "out of place" or unsympathetic to the design, I think there is room for some styling consolidation as part of the next iteration. I'd be interested to know your thoughts on if there is existing styling guidelines for matching the exact native platform colors when re-implementing platform-specific behavior in Javascript. I think it would be an interesting conversation to start, if not.

fluiddot replied:
Great investigation @derekblank! I agree that none of the examples look incoherent in terms of styling, so I'd stick with your approach. However, I'd like to ping here @iamthomasbishop, in case he could provide further insights regarding this topic.

@iamthomasbishop I'd appreciate it if you could take a look at this and confirm the right colors for the clear input button, thank you very much for the help 🙇 !

fluiddot commented:
> Thanks for the feedback @fluiddot ! After considering your comments, and thinking through how we might resolve some of the edge cases like retaining the `clipboardUrl` in LinkPicker, I realized that a better approach would be to provide an `onClear` method as a prop to BottomSheet.Cell. This way, each use of `displayClearButton` in BottomSheet.Cell would be accompanied by its own `onClear` handler that the clear button itself could invoke, so that the TextInput value can be handled directly in the parent component of wherever BottomSheet.Cell is being used:
> 
> ```
> const clearInput = () => {
>     setText( '' );
> };
> 
> <BottomSheet.Cell
>     displayClearButton={true}
>     onClear={() => clearInput()}
>     value={text}
>     ...
> />
> ```
> 
> It is an extra prop and function, yes, but I think it is much cleaner as it eliminates the need for a `localValue` state within BottomSheet.Cell, and it allows the parent component to maintain control of the value and provide extra parameters like `setValue( { value, clipboardUrl: '' } )` when needed. In addition, this change resolves several of your other comments on the code, but that feedback was helpful. Let me know your thoughts on this approach! 🙇

Good job on addressing the issue 🎊 . I like the approach of delegating the clear input functionality to the invoker, this way we make the component agnostic to the data management.

derekblank committed:
Transfer clearIcon margin style to touchable container

fluiddot reviewed (commented):
@derekblank I've tested the changes in both iOS and Android and identified the following issues, none of them should block the PR but I think it would be great to try addressing them:

### Android - Visual glitch
I noticed that on Android the content of empty text inputs is aligned to left and when introducing a value they change the alignment to right. This is a visual glitch that is present in the current version so it's not produced by these changes, however, in combination with displaying the clear button, it makes it more prominent leading to a minor visual glitch (see attached video captures).

<details><summary>Click here to display video captures</summary>

Current version|Including clear button
--|--
<video src=https://user-images.githubusercontent.com/14905380/147585354-daff5712-2eeb-4dc0-9aad-ade69f62709d.mp4>|<video src=https://user-images.githubusercontent.com/14905380/147585380-04efa1b5-d050-43fa-b5e6-00ddd0c12ac4.mp4>

</details>

### Content size reduced by clear button
In some cases where in the same row we present the text input along with an icon and text label, like for example when editing the URL of an embed block, the available size for the content is reduced when the clear button is displayed resulting in a small area (see attached screenshot).

<details><summary>Click here to display screenshot</summary>

<img src=https://user-images.githubusercontent.com/14905380/147586672-8e2ebe47-2e19-428c-85f8-e118c3e83856.jpg width=400>

</details>

This is more a design issue than implementation but I'm wondering if we could figure out a solution eventually until further design feedback is provided, wdyt?

cc @iamthomasbishop 

### Enable clear button in Link rel input
While I was testing the Image block, I noticed that the Link Rel input could be benefited from the clear input button functionality (see attached screenshot) as it's a similar input to the Link label used on other blocks. I'm wondering if we could enable it there too, wdyt?

<details><summary>Click here to display screenshot</summary>

<img src=https://user-images.githubusercontent.com/14905380/147587316-56601287-30c6-4d83-b743-a0d47aa7e353.jpg width=400>

</details>

<hr>

Apart from the mentioned issues, the PR looks great to me 🎊 . As the last step and since the new functionality is part of the Cell component which is a core piece of the editor, I'm wondering if we could add some unit tests to cover the changes introduced in the PR.

derekblank committed:
Add clear button behavior to LinkSettings Link Rel input

derekblank commented:
@fluiddot Thanks for the thorough review! These are some good points. Per the last point:

> ### Enable clear button in Link rel input
> While I was testing the Image block, I noticed that the Link Rel input could be benefited from the clear input button functionality (see attached screenshot) as it's a similar input to the Link label used on other blocks. I'm wondering if we could enable it there too, wdyt?

This is a good callout. I've went ahead and updated this component to add the clear button behavior to Link Rel on LinkSettings.

Regarding the other two points (the Android visual glitch and the content size design issue), I have done some further investigations here on these two specific issues that I will post my thoughts on these here shortly.  Also, I agree with you that the next steps would be to add some basic unit testing for clear button behavior on the BottomSheet Cell component.

fluiddot commented:
> This is a good callout. I've went ahead and updated this component to add the clear button behavior to Link Rel on LinkSettings.

Awesome! Thanks for applying the suggestion 🙇 .

> Regarding the other two points (the Android visual glitch and the content size design issue), I have done some further investigations here on these two specific issues that I will post my thoughts on these here shortly

Great, I look forward to your investigations 🔍 .

<hr>

@derekblank I created a test PR in the Gutenberg Mobile repository ([reference](https://github.com/wordpress-mobile/gutenberg-mobile/pull/4415)) in order to verify that E2E tests and CI checks are passing. It would be great if you could take a look in case any of them fail and require further fixing, thanks!

derekblank commented:
@fluiddot Just following up with updates to the other two of your comments: 

### 1. Android - Visual glitch
> I noticed that on Android the content of empty text inputs is aligned to left and when introducing a value they change the alignment to right. This is a visual glitch that is present in the current version so it's not produced by these changes, however, in combination with displaying the clear button, it makes it more prominent leading to a minor visual glitch (see attached video captures).

This is a good callout, but it is also interesting because when focusing the TextInput on my testing device, the cursor appears to the **right** of the placeholder text by default, unlike your example where the cursor appears on the **left** of the TextInput before any text is actually added. This may be an OS-level difference between Android devices or Android versions -- I'm not entirely sure at this point without investigating further with different test devices. (I am using a Google Pixel 6 on Android 12, for reference.) I know that it is possible with React Native's TextInput to modify the start and end value of the text selection's position using the [selection](https://reactnative.dev/docs/textinput#selection) property, which may help in this case. I am curious which device and OS you are using to reproduce this, however. Regardless, I experienced the same glitchy behavior you mentioned, even with the cursor position already set to the "right" of the TextInput placeholder...so while it is certainly an issue, I'm not entirely certain that it is caused by the right-vs-left position of the cursor. Investigating further, I did a 60fps screen recording of an interaction with the TextInput, and broke out every frame into a separate image file using `ffmpeg`. Here's an example of the output:

![input focus 2](https://user-images.githubusercontent.com/643285/147808849-088a55ef-1dd0-4929-8217-4ae100001503.png)

What I interpreted was that between Frame 8 and Frame 11, there was a three frame gap where the actual text in the text input appeared, but the clear button did not appear until three frames later. (The gap from Frame 1 to Frame 8 was just me moving my finger from the input to the keyboard.) Without a bit more exploration, my best guess as to why this is the case is because the presence of the clear button in BottomSheet.Cell is checking for the existence of the TextInput `value`, which must be passed up to the parent component and then back down again to BottomSheet.Cell.:

```
{ value !== '' && displayClearButton && clearButton() }
```

This behavior only shows the clear button when the TextInput actually has some value, but it also leads to the "glitchy" behavior that is more noticeable on right-aligned TextInputs. Looking into the behavior further of what else we might be able to hook into to display the clear button, I tried changing the check for any TextInput `value` to local state when the input is focused, like this:

```
{this.state.isEditingValue && displayClearButton && clearButton() }
```

This now displays the clear input button whenever the text input is **focused**, regardless of when the TextInput actually has value. This reduces the glitchy behavior (especially on right-aligned TextInputs). Here's an example in action:

<details>
<summary>Click to display persistent Clear Button via `this.state.isEditingValue`</summary>

https://user-images.githubusercontent.com/643285/147737912-bc494f4f-1904-44b9-8342-782a9e3e393f.mov

</details>

So, while the clear button displays immediately on TextInput focus without any TextInput value, and it reduces the glitchy behavior, this differs a bit from the previous behavior where the clear button only appears once value has been entered. I did not push up this change, as I wanted to discuss and get some more input. I think we could make a case for moving to this behavior, as there is a native precedent for this on iOS on the clearButtonMode [`while-editing`](https://developer.apple.com/documentation/uikit/uitextfield/viewmode/whileediting) property. Another alternative would be to continue investing ways to speed up the `value !== ''` check, or some other creative ways to make the behavior appear less glitchy in the UI (since it mostly affects right aligned inputs). What do you think? 

### 2. Content size reduced by clear button
> In some cases where in the same row we present the text input along with an icon and text label, like for example when editing the URL of an embed block, the available size for the content is reduced when the clear button is displayed resulting in a small area (see attached screenshot).

I agree that it is not ideal that the clear button takes up even more space here. One possible implementation solution would be to reduce the left margin of the clear button container for 16 to 12, to match the text block on the left side. That's not much gain, of course, but it is a small amount. Another option that would be more of a design consideration would be to explore moving these text blocks out of the input row to free up more horizontal space -- perhaps above, below, or even explore placeholder text options as well? Just some initial ideas.

fluiddot commented:
@derekblank Thanks for investigating both issues, you did a superb investigation 🙇 .

> 1. Android - Visual glitch

> This is a good callout, but it is also interesting because when focusing the TextInput on my testing device, the cursor appears to the right of the placeholder text by default, unlike your example where the cursor appears on the left of the TextInput before any text is actually added. This may be an OS-level difference between Android devices or Android versions -- I'm not entirely sure at this point without investigating further with different test devices. (I am using a Google Pixel 6 on Android 12, for reference.) I know that it is possible with React Native's TextInput to modify the start and end value of the text selection's position using the selection property, which may help in this case. I am curious which device and OS you are using to reproduce this

Interesting, I didn't realize that this might be different depending on the device/OS version, in my case I'm using the device Samsung Galaxy S20 FE 5G (Android 11).

Not sure if the `selection` prop would help since I only managed to reproduce this when the input is empty. As you mentioned, this might be an issue on the Android side that doesn't preserve alignment for the placeholder element 🤔 .

> Regardless, I experienced the same glitchy behavior you mentioned, even with the cursor position already set to the "right" of the TextInput placeholder...so while it is certainly an issue, I'm not entirely certain that it is caused by the right-vs-left position of the cursor. Investigating further, I did a 60fps screen recording of an interaction with the TextInput, and broke out every frame into a separate image file using `ffmpeg`. Here's an example of the output:
> 
> ![input focus 2](https://user-images.githubusercontent.com/643285/147808849-088a55ef-1dd0-4929-8217-4ae100001503.png)
> 
> What I interpreted was that between Frame 8 and Frame 11, there was a three frame gap where the actual text in the text input appeared, but the clear button did not appear until three frames later. (The gap from Frame 1 to Frame 8 was just me moving my finger from the input to the keyboard.) Without a bit more exploration, my best guess as to why this is the case is because the presence of the clear button in BottomSheet.Cell is checking for the existence of the TextInput `value`, which must be passed up to the parent component and then back down again to BottomSheet.Cell.:
> 
> ```
> { value !== '' && displayClearButton && clearButton() }
> ```
> 
> This behavior only shows the clear button when the TextInput actually has some value, but it also leads to the "glitchy" behavior that is more noticeable on right-aligned TextInputs. Looking into the behavior further of what else we might be able to hook into to display the clear button, I tried changing the check for any TextInput `value` to local state when the input is focused, like this:
> 
> ```
> {this.state.isEditingValue && displayClearButton && clearButton() }
> ```
> 
> This now displays the clear input button whenever the text input is **focused**, regardless of when the TextInput actually has value. This reduces the glitchy behavior (especially on right-aligned TextInputs).

Great findings, I agree that this looks related to the fact that it requires multiple renders to determine whether to display the clear button, leading to the visual glitch. As you commented, this is prominent only in the right-aligned text inputs, so I'd focus on fixing that case. The approach about keeping the clear button even when it's empty would be a good workaround, however, I'm thinking of only applying it for the right-aligned case, wdyt?

> ### 2. Content size reduced by clear button
> I agree that it is not ideal that the clear button takes up even more space here. One possible implementation solution would be to reduce the left margin of the clear button container for 16 to 12, to match the text block on the left side. That's not much gain, of course, but it is a small amount. Another option that would be more of a design consideration would be to explore moving these text blocks out of the input row to free up more horizontal space -- perhaps above, below, or even explore placeholder text options as well? Just some initial ideas.

I like the idea of reducing it even though it's a small amount, keep in mind that if we make it too small we might re-introduce the issue about overlapping the button with text selection ([reference](https://github.com/WordPress/gutenberg/pull/37402#discussion_r769856902)). Nevertheless, I'd like to know the feedback from @iamthomasbishop here, in case we could make any design tweak in order to improve tex inputs.

derekblank committed:
Add basic tests for clear button behavior in BottomSheet Cell

fluiddot commented on packages/components/src/mobile/bottom-sheet/cell.native.js:
```diff
@@ -263,6 +264,7 @@ class BottomSheetCell extends Component {
 						numberOfLines={ 1 }
 						style={ finalStyle }
 						value={ value }
+						testID="text-input"
```
I think you could query for the text input without requiring adding a test ID, the React Native testing library provides the `ByDisplayValue` queries that should fit your case ([reference](https://callstack.github.io/react-native-testing-library/docs/api-queries#bydisplayvalue)).

fluiddot replied:
Thanks for addressing this in https://github.com/WordPress/gutenberg/pull/37402/commits/62b914475964edbf26d5e817d2f0bcff3e3ec903 🙇 .

fluiddot commented on packages/components/src/mobile/bottom-sheet/cell.native.js:
```diff
@@ -369,6 +371,7 @@ class BottomSheetCell extends Component {
 				onPress={ onCellPress }
 				onLongPress={ onLongPress }
 				style={ [ styles.clipToBounds, style ] }
+				testID={ 'cell-touchable' }
```
I see this component already defines an accessibility label so you could use it for querying it instead of adding a test ID. The React Native testing library provides queries `ByA11yLabel` for this case ([reference](https://callstack.github.io/react-native-testing-library/docs/api-queries#bya11ylabel-byaccessibilitylabel-bylabeltext)).

derekblank replied:
This was one of the area where I thought having a dedicated testID might be advantageous, as this component's a11yLabel is really conditional. But, for the clear button tests we're trying to scope, I can also see how using the accessibility label might be more realistic here, so I will consider that as part of updating the tests.

fluiddot replied:
For the test case, one thing you could try is to pass a value through in the `accessibilityLabel` prop when rendering the `BottomSheetCell` component, this way you could use that value for querying it.

derekblank replied:
Ah, yep! I did not even consider that it would be just as easy to pass the accessibilityLabel as a prop here for testing. Thanks for this suggestion!

fluiddot commented on packages/components/src/mobile/bottom-sheet/cell.native.js:
```diff
@@ -199,6 +199,7 @@ class BottomSheetCell extends Component {
 				<TouchableOpacity
 					onPress={ onClear }
 					style={ styles.clearIconContainer }
+					testID="clear-button"
```
Instead of adding a test ID, I'm wondering if we could set accessibility properties to this component, as we do for the main button, in fact, this would help us improve the accessibility of the editor. Wdyt?

https://github.com/WordPress/gutenberg/blob/666e9ea5d18eed5a337c0d21faa7b356bfa4dce0/packages/components/src/mobile/bottom-sheet/cell.native.js#L355-L368

derekblank replied:
Good callout! I've update the test and accessibility labels here. The one part I am curious on is if I have handled the i18n effectively here. Tracing back the thread to the [i18n __( ... )](https://github.com/WordPress/gutenberg/blob/52bac1c1d524b333c19872a07854d7e7121dc337/packages/i18n/README.md#__) method, the documentation suggests this is ok to just use a text string here. Let me know if this needs further investigation.

fluiddot replied:
> Tracing back the thread to the i18n __( ... ) method, the documentation suggests this is ok to just use a text string here. Let me know if this needs further investigation.

Yeah, the usage of the i18n function is right. We don't need to take further actions regarding translations, as this is automatically handled by the translation pipeline.

fluiddot commented on packages/components/src/mobile/bottom-sheet/test/cell.native.js:
```diff
@@ -0,0 +1,102 @@
+		expect( clearButton ).toBeNull();
+	});
+
+	it( 'clears text input value when the clear button is pressed', async () => {
```
This test is checking if the handler is called, which doesn't guarantee that the text input would be actually cleared. I'm wondering if we should update the test's description in order to be more accurate regarding what it's covering, wdyt?

derekblank replied:
Good call! I've updated the test description here to reflect that only the `onClear` handler will be called.

fluiddot replied:
Great, thanks for updating it 🙇 .

fluiddot commented on packages/components/src/mobile/bottom-sheet/test/cell.native.js:
```diff
@@ -0,0 +1,102 @@
+		const cellTouchable = await screen.getByTestId( 'cell-touchable' );
+		fireEvent.press( cellTouchable );
+
+		// Simulate onChangeValue events (without rerender)
+		const textInput = await screen.getByTestId( 'text-input' );
+		fireEvent( textInput, 'focus' );
+		fireEvent.changeText( textInput, newTestValueString );
+		expect( onChangeValueMock ).toHaveBeenCalledWith( newTestValueString );
```
I understand that this test verifies that the clear button is displayed when it has text, however, I see that it also changes the text, is it really required? I'm wondering if it would be worth it to split it into two tests:
1. Check that the clear button is displayed when the text has value.
2. Check that the clear button remains displayed when the text has value and changes. [Although this seems a bit redundant, wdyt?]

derekblank replied:
I agree with your reasoning here. Initially when I wrote these few lines of this test, I was expecting to use `rerender`, but I agree that this is redundant under a scoped component approach so I've removed these lines.

fluiddot replied:
Thanks for updating the tests with this suggestion 🙇 .

fluiddot reviewed (commented):
@derekblank I've just performed a review on the last changes, awesome work on adding the test cases for the Cell component 🥇 .

I added some comments regarding the addition of test IDs, let me know if you could take a look at them, thanks 🙇 !

fluiddot commented:
@derekblank I noticed that the PR checks are not being triggered due to file conflicts, it would be great if you could solve them in order to make them pass, thanks 🙇 !

derekblank committed:
Update code style

- Removes unused  from link-picker/index.native.js
   (which was retained as the result of a merge conflict)
- Fixes Prettier style update

derekblank committed:
Update accessibility labels and test IDs for BottomSheet Cell

fluiddot commented on packages/components/src/mobile/bottom-sheet/cell.native.js:
```diff
@@ -196,7 +196,10 @@ class BottomSheetCell extends Component {
-					testID="clear-button"
+					accessibilityLabel={ __( 'Clear Button' ) }
+					accessibilityHint={ __(
+						'Will clear the value of the text input when pressed'
```
I checked other hint messages ([reference](https://github.com/WordPress/gutenberg/blob/a8061c1db1e4252fb14d0ee033cf3340d694b114/packages/block-library/src/missing/edit.native.js#L91)) and I noticed that we also include the action to be made by the user, so I'm wondering if we should update the message with something like:
```Tap here to clear the value of the text input```

What do you think?

derekblank replied:
I think it's a great idea -- I've updated the accessibility hint to use the imperative mood.

derekblank committed:
Update BottomSheet Cell tests to prefer a11yLabels as test selectors

derekblank committed:
Add clear button  integration test to LinkPicker

derekblank commented on packages/components/src/mobile/bottom-sheet/test/cell.native.js:
```diff
@@ -22,14 +24,15 @@ describe( '<BottomSheetCell />', () => {
 
 		const screen = render(
 			<BottomSheetCell
+				accessibilityLabel={ 'Text Input' }
```
@fluiddot Ah, yep! I did not even consider that it would be just as easy to pass the a11yLabel as a prop here for testing. (Previously I was using testID.) Thanks for this suggestion!

iamthomasbishop commented:
@derekblank Thank you for tackling this, it's looking pretty solid. In the long-term, I would prefer to move to a more scalable solution for BottomSheet Text Inputs (more on that below) but I think we will be in good shape if we can solve the "jumpiness" that happens when focusing and adding text on right-aligned inputs. 

I put together a quick Figma prototype to see if applying a quick/subtle transition would help and it does seem to feel a little more natural. I think I would prefer this solution over always showing the `clear` icon/button because that diverges from what is an established pattern. Here is an example using a quick (50ms) `easeOut` curve to change the x-position of the text label upon text-entry:

https://user-images.githubusercontent.com/1200432/148587346-8e08da30-3c87-4164-9e7e-9bfeebc65a03.mov

## For the future

Note: This is probably beyond scope for this project, but it is related so I wanted to write it down so that it's documented.

As I mentioned above, I would prefer that we move to a more scalable model in general for text-entry — esp within BottomSheets. For context, we do already have something roughly equivalent to a `textarea` that we're using for alt-text settings on Image blocks but I think we want something more compact in most cases that resides within a single-line cell. Here are a couple of options I have in mind:

**Option A** (preferred)

The text-entry task is performed on a sub-sheet, which is accessed by tapping on the cell label/disclosure. This provides more space for the Text Input itself, but also allows us to provide additional information/contextual elements to users if necessary.

<a href="https://user-images.githubusercontent.com/1200432/148588372-010c3e27-2312-41cc-a3ed-df39ff3f0185.png"><img src="https://user-images.githubusercontent.com/1200432/148588372-010c3e27-2312-41cc-a3ed-df39ff3f0185.png" width="1000"></a>


**Option B**

We maintain the same inline text-entry, but left-align the value/right side of the cell. This would be somewhat consistent with  the ["Left Detail" variation](https://developer.apple.com/design/human-interface-guidelines/ios/views/tables/) ([`UITableViewCellStyleValue2`](https://developer.apple.com/documentation/uikit/uitableviewcellstyle/uitableviewcellstylevalue2)) from the Human Interface Guidelines, but would likely cause some awkward visual alignment issues. I would definitely prefer **Option A** over this.

<a href="https://user-images.githubusercontent.com/1200432/148589791-d9a83970-c649-45bb-83c7-e0efd63216be.png"><img src="https://user-images.githubusercontent.com/1200432/148589791-d9a83970-c649-45bb-83c7-e0efd63216be.png" width="1000"></a>

derekblank commented:
@iamthomasbishop Thanks for putting together the animation focusing on the jumpiness. I think this is a good place to start! My initial concern was that three frame gap between when the text input value appears, and when the logic of displaying the clear button can complete, so if we can't improve this, I think we can use the animation to help mask it. 

Regarding the future implementation, I very much agree with your Option A plan -- this seems like a good approach to free up more horizontal space. As a side note, we'll be hitting pause on this work for the time being, but I would love to resume discussing this implementation later on.

ttahmouch commented on packages/components/src/mobile/link-settings/test/edit.native.js:
```diff
@@ -428,5 +428,147 @@ describe.each( [
+			 * GIVEN the LINK PICKER field is FOCUSED but has NO text value;
+			 */
+			// eslint-disable-next-line jest/no-done-callback
+			it( 'should not display a CLEAR BUTTON within the LINK SETTINGS field', async ( done ) => {
```
**Not Crazy Important:** The `done` callback usages will likely need to be removed when you update your branch because of recent changes related to `jest-circus`.
https://github.com/WordPress/gutenberg/pull/37715/files#r780551266

derekblank replied:
@ttahmouch Thanks for the heads up! 👍
